### PR TITLE
[NNC] Fixes case where inlining wouldn't work because dim-size was 1.

### LINKED ISSUE #52581: [TensorExpr] Loop transformations: `compute_inline` sometime fails when dim-size is 1
## 🐛 Bug

`compute_inline` is a transformation that tries to replace a load from a buffer with its defining expression. In order to find this expression, it scans the TE statement looking for stores to that buffer - it expects to find a single store with simple indices. Simple indices here means that they are just `Var`s and not compound expressions (this is needed so that we can figure out to correctly rewrite the load).

The problem arises when some of the dimensions of the buffer have size 1 and the statement has been simplified: in this case indices that previously were `Var`s could be rewritten with `0`. In this case we still can unambiguously recreate the expression for the inlined buffer, so we should do it.

## To Reproduce
Consider the following script:
```python
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

with kernel_arena_scope():
    dtype = torch._C._te.Dtype.Float
    M = 1
    K = 9
    N = 1

    def get_dim_args(dims):
        dim_args = []
        for dim in dims:
            dim_args.append(torch._C._te.DimArg(dim, 'i' + str(len(dim_args))))
        return dim_args

    (MM, KK, NN) = [torch._C._te.ExprHandle.int(x) for x in [M, K, N]]

    dtype = torch._C._te.Dtype.Float
    X = torch._C._te.Placeholder('X', dtype, [MM, KK, NN])
    Y = torch._C._te.Compute('Y', get_dim_args([MM, KK, NN]),
            lambda m, n, k: X.load([m, n, k]))
    Z = torch._C._te.Compute('Z', get_dim_args([MM, KK, NN]),
            lambda m, n, k: Y.load([m, n, k]))
    loopnest = torch._C._te.LoopNest([Z])
    print('Original statement:', loopnest.root_stmt())

    # If we un-comment this 'simplify', compute_inline will fail
    # loopnest.simplify()
    # print('After simplify:', loopnest.root_stmt())

    loopnest.compute_inline(loopnest.get_loop_body_for(Y))
    print('After compute_inline(Y):', loopnest.root_stmt())
```

If we run it as-is, `compute_inline` will succeed, as expected. The output will be:
```
Original statement: {
  for (int i0 = 0; i0 < 1; i0++) {
    for (int i1 = 0; i1 < 9; i1++) {
      for (int i2 = 0; i2 < 1; i2++) {
        Y[i0, i1, i2] = X[i0, i1, i2];
      }
    }
  }
  for (int i0_1 = 0; i0_1 < 1; i0_1++) {
    for (int i1_1 = 0; i1_1 < 9; i1_1++) {
      for (int i2_1 = 0; i2_1 < 1; i2_1++) {
        Z[i0_1, i1_1, i2_1] = Y(i0_1, i1_1, i2_1);
      }
    }
  }
}
After compute_inline(Y): {
  for (int i0 = 0; i0 < 1; i0++) {
    for (int i1 = 0; i1 < 9; i1++) {
      for (int i2 = 0; i2 < 1; i2++) {
        Z[i0, i1, i2] = X[i0, i1, i2];
      }
    }
  }
}
```

However, if we run IRSimplifier before `compute_inline`, it will fail:
```
After simplify: {
  for (int i1 = 0; i1 < 9; i1++) {
    Y[0, i1, 0] = X[0, i1, 0];
  }
  for (int i1_1 = 0; i1_1 < 9; i1_1++) {
    Z[0, i1_1, 0] = Y(0, i1_1, 0);
  }
}
Traceback (most recent call last):
  File "q.py", line 37, in <module>
    loopnest.compute_inline(loopnest.get_loop_body_for(Y))
RuntimeError: cannot inline Buf with compound indices
```

## Expected behavior
`compute_inline` works on the simplified stmt:
```
After simplify: {
  for (int i1 = 0; i1 < 9; i1++) {
    Y[0, i1, 0] = X[0, i1, 0];
  }
  for (int i1_1 = 0; i1_1 < 9; i1_1++) {
    Z[0, i1_1, 0] = Y(0, i1_1, 0);
  }
}
After compute_inline(Y): {
  for (int i1_1 = 0; i1_1 < 9; i1_1++) {
    Z[0, i1_1, 0] = X[0, i1_1, 0];
  }
}
```

### PR DESCRIPTION
**Pull request by Chillee**
Fixes https://github.com/pytorch/pytorch/issues/52581

The git diff is absolutely atrocious since I also refactored the code to share stuff between `Load` and `FunctionCall`.

Biggest questions I have about this diff are:

1. The asserts I added. From my understanding it's not possible to have a constant index in `Store` that's non-zero, since `Store` always creates a new buffer. Perhaps the user can write this kind of incorrect code, though, so perhaps I should just check for it and not assert it?

2. I don't think(?) I need to do any special handling for `index_vars`, but wasn't totally able to track the logic there.

### CODE REVIEW

**Review by bertmaher — CHANGES_REQUESTED** [review]
Could you add a test case for this?  It'd help me understand the problem it's solving a bit better.

**Comment by Chillee** [comment]
@bertmaher I added a test case but I think Mikhail's example in https://github.com/pytorch/pytorch/issues/52581 is more clear.

**Review by bertmaher — APPROVED** [review]

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 1]
```diff
@@ -533,34 +533,38 @@ class FunctionInliner : public IRMutator {
...
-    if (buf != buf_) {
-      return IRMutator::mutate(v);
-    }
-
-    if (v->nparams() != buf->ndim()) {
-      throw malformed_input(
-          "Placeholder indexed access is inconsistent with its rank", v);
-    }
-
+  const Expr* mutate_loads(const Buf* buf, std::vector<const Expr*> dims) {
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:553]: Kind of a tip to make future diffs like this easier to review, it's nice to stack changes that include refactoring and functionality changes as two diffs: (1) refactor, (2) new behavior.  That way I can glance through (1) to verify it's just code motion, and then use my actual brainpower (such as it is) on the functional changes.

If you use ghstack it makes the stacked diff experience basically as nice as phabricator :-)

Comment (Chillee) [reply torch/csrc/jit/tensorexpr/loopnest.cpp:553]: right, I'll switch to ghstack this weekend :P

**Pull request closed without merging**